src/main/java/cz/cvut/kbss/study/service/repository/RepositoryUserService.java

```diff
@@ -202,8 +202,8 @@ protected void prePersist(User instance) {
     @Override
     protected void preUpdate(User instance) {
         final User currentUser = securityUtils.getCurrentUser();
-        if (!currentUser.getTypes().contains(Vocabulary.s_c_administrator)
-            && (!instance.getTypes().equals(currentUser.getTypes()) || (instance.getInstitution() != null
+        if (!currentUser.isAdmin()
+            && (!instance.getRoleGroup().getRoles().equals(currentUser.getRoleGroup().getRoles()) || (instance.getInstitution() != null
```

Comment:
Wow, not sure if I understand role of this condition now or before, but I am little worried that we need compare roleGroups here instead, not roles itself as we might persist different role of user.

Reply:
? (it seems to me that this code is used for both keycloak-auth and internal-auth)

Reply:
@palagdan

Reply:
@palagdan please answer this

Reply:
The condition is true when:
- The user is not an admin, and
- Either the roles of the instance and the currentUser are different, or:
- The instance has an institution, and its key is different from the currentUser's institution key.

You are right, compare the role groups will be more correct.